### PR TITLE
Feature/isilon size and preemption

### PR DESCRIPTION
**Pull request by danforrest**
- Added support for the --size flag
- Added the dataSubnet configuration parameter
- Added support for pre-emption
- Updated documentation

The instance ID will be a concatenation of all IP addresses that may connect to an attached volume.  All IPs on the host, that are also in the data subnet, will be allowed to attach to the volume.

### CODE REVIEW

**Comment by clintonskitson** [comment]
@danforrest Thanks for the updates!

A few things

1) You want to squash these commits into a single update.  `git rebase -i HEAD~7`.
2) Make sure to run `go fmt` to standardize the formatting of your code.
3) The glide.yaml file needs to be updated to point at your update goisilon package for it to compile correctly.  You should probably open a PR again the goisilon repo first, and then merge it.  Then the commit and repo to point to would be correct.

**Comment by clintonskitson** [comment]
I see this was re-pushed.  The way to properly bring commits to the head or make up for differences in updates to the upstream branch is a rebase.  Typically you would rebase your commits onto the latest head of master.  But in your case, I would first reset to the head and just cherry pick your changes in.
1) Get latest commits from emccode branch `git fetch -u emccode`
2) `git reset --hard ourlatestcommit`
3) `git cherry-pick 5cfa`
4) `git cherry-pick fcbd`

Following this you should `rebase -i HEAD~2` to bring your changes into a single commit.  I still can't compile it since the depedent go-isilon package doesn't seem to be at the commit it needs to be.

**Comment by clintonskitson** [comment]
I see this was re-pushed.  The way to properly bring commits to the head or make up for differences in updates to the upstream branch is a rebase.  Typically you would rebase your commits onto the latest head of master.  But in your case, I would first reset to the head and just cherry pick your changes in.
1) Get latest commits from emccode branch `git fetch -u emccode`
2) `git reset --hard ourlatestcommit`
3) `git cherry-pick 5cfa`
4) `git cherry-pick fcbd`

Following this you should `rebase -i HEAD~2` to bring your changes into a single commit.  I still can't compile it since the depedent go-isilon package doesn't seem to be at the commit it needs to be.

**Comment by clintonskitson** [comment]
I see this was re-pushed.  The way to properly bring commits to the head or make up for differences in updates to the upstream branch is a rebase.  Typically you would rebase your commits onto the latest head of master.  But in your case, I would first reset to the head and just cherry pick your changes in.
1) Get latest commits from emccode branch `git fetch -u emccode`
2) `git reset --hard ourlatestcommit`
3) `git cherry-pick 5cfa`
4) `git cherry-pick fcbd`

Following this you should `rebase -i HEAD~2` to bring your changes into a single commit.  I still can't compile it since the depedent go-isilon package doesn't seem to be at the commit it needs to be.

**Comment by akutz** [comment]
Hi @danforrest,

Per @clintonskitson's feedback we're still waiting for the requested updates prior to merging this commit. We're close to ready to release 0.3.1, and if you're wanting these changes in that release we'd ask you to go ahead and squash things down and reduce this PR to a single commit since it really seems to be a single commit is all that's necessary. Additionally, please see the [REX-Ray dev guide](http://rexray.readthedocs.org/en/latest/dev-guide/project-guidelines/) for rules regarding commit messages. Thanks!

**Comment by akutz** [comment]
Hi @clintonskitson,

So what's the word on this? Omit from 0.3.1? @danforrest has yet to respond.

**Comment by clintonskitson** [comment]
Yes I believe we can untag this one for now.

**Comment by danforrest** [comment]
Sorry for the delay, I was on vacation for the past couple weeks.  I've rebased the commit per the comment from @clintonskitson.  As for the GoIsilon package, the correct commit is 336be8, and that should be the one in glide.yaml.  Is it not able to find this commit or does it fail when compiling?

File: .docs/user-guide/storage-providers/isilon.md [hunk 1]
```diff
@@ -19,6 +19,7 @@ isilon:
   password: password
   volumePath: /rexray
   nfsHost: nfsHost
+  dataSubnet: subnet
```
Comment (clintonskitson) [inline .docs/user-guide/storage-providers/isilon.md]: Is the parameter supposed to be `dataSubnets`?

**Comment by clintonskitson** [comment]
Can you send me over a SmartQuotas license for testing?

**Comment by clintonskitson** [comment]
I am not seeing the size populate when doing a `volume ls` for volumes created.  Is this expected without the quota turned on?

**Comment by clintonskitson** [comment]
Should there be verbage added to the docs around the requirements to make the capacity settings function?

File: .docs/user-guide/storage-providers/isilon.md (hunk 1 above)
Comment (danforrest) [inline .docs/user-guide/storage-providers/isilon.md]: Yep.  Updated all occurrences to be dataSubnets.

**Comment by danforrest** [comment]
Added documentation mentioning the need for a SmartQuota license to make the capacity settings function.

I'm looking into getting you a license for test purposes now as well.

**Comment by danforrest** [comment]
@clintonskitson from what I can find, license requests for the Isilon Simulator need to go through isilon.community@emc.com or a service request.

If that doesn't work, you can try isi-licenses@isilon.com, but I believe that is for normal paid licenses.

**Comment by danforrest** [comment]
You can also try 'isi_get_license' from the command line on the cluster.  I know it works for getting evaluation licenses on the Isilon network.  It may also work for clusters on any EMC network, but I haven't tried it yet.

**Comment by clintonskitson** [comment]
Haven't applied license yet.  Getting this error by default now.

```
[root@localhost temp]# ./rexray volume create --volumename=testing1 --size=1
FATA[0000] error creating volume                         inner=Unable to set acl -- operation not permitted. provider=Isilon volumeName=testing1
```

**Comment by clintonskitson** [comment]
See https://github.com/clintonskitson/rexray/commit/355372d404a5cabd2604fcb22e223496767fca8c for example of how to properly do debug logging and error logging using framework.  I was getting nil exception errors because there was no error checking for before it got to trying to retrieve the volume that was created.

File: .docs/user-guide/storage-providers/isilon.md (hunk 1 above)
Comment (clintonskitson) [inline .docs/user-guide/storage-providers/isilon.md]: Should this be dataSubnets or dataSubnet?

**Comment by clintonskitson** [comment]
I activated the license, and am still getting the same error from above.

**Comment by danforrest** [comment]
I think I see what's causing the ACL error.  I forgot to go back and remove a hard coded user/group in the CreateVolume method.  I'll fix it in the goisilon code.

**Comment by clintonskitson** [comment]
Also, I would assume that this wouldn't impact someone if they aren't using quotas.  Maybe a flag in the config to enable quotas/size?

**Comment by danforrest** [comment]
Fixed the permissions issues.  The owner of a new volume is now the user specified in the config file rather than `nobody`.  Also, this user must have these permissions on the cluster: ISI_PRIV_NS_IFS_ACCESS, ISI_PRIV_LOGIN_PAPI, ISI_PRIV_NFS, ISI_PRIV_QUOTA, and ISI_PRIV_IFS_RESTORE.

Added the config parameter `quotas` to enable/disable capacity functionality.

I'll handle squashing the commits after we get all the bugs ironed out.

File: .docs/user-guide/storage-providers/isilon.md [hunk 2]
```diff
@@ -47,15 +51,26 @@ isilon:
...
+It is expected that the `volumePath` exists already within the Isilon system.  This would reflect a directory create under `/ifs/volumes/rexray`.  It is not necessary to export this volume.  The `dataSubnets` parameter is required so the Isilon driver can restrict access to attached volumes to the host that REX-Ray is running on.
+
+If `quotas` are enabled, a SmartQuotas license must also be enabled on the Isilon cluster for the capacity size functionality of REX-Ray to work.
 
 ## Caveats
 
-- This driver currently ignores the `--size` and `--volumeType` flags.
-- This driver does not support pre-emption currently.  Requests from alternate hosts can have bad results.
+- This driver currently ignores the `--volumeType` flag.
+- The account used to access the Isilon cluster must have these privaleges:
```
Comment (clintonskitson) [inline .docs/user-guide/storage-providers/isilon.md]: Typo in privileges here

**Comment by clintonskitson** [comment]
Thanks @danforrest, getting there!  Tested good from my end. 
1) Let's make `dataSubnets` to be `dataSubnet`, unless it can actually take multiple subnets.
2) Can you make sure parameters like group are defined in the README?
3) Let's squash all of the commits into a single one

**Comment by clintonskitson** [comment]
Btw.  Try and make sure you adhere to these [guidelines](http://rexray.readthedocs.org/en/stable/dev-guide/project-guidelines/) for commit messages.  Your message above re groups was unreadable.

File: .docs/user-guide/storage-providers/isilon.md (hunk 2 above)
Comment (danforrest) [inline .docs/user-guide/storage-providers/isilon.md]: Fixed.

**Comment by danforrest** [comment]
- Changed dataSubnets to dataSubnet in documentation.
- Added parameters to GoIsilon README.
- Updated glide.yaml to point to merged GoIsilon commit.
- Squashed commits.

**Comment by clintonskitson** [comment]
@danforrest I just spotted a problem with the formatting of the driver file.  Can you run `make test` and then commit the changes?  This should kick off `go fmt` to ensure we don't have unneeded diffs.

**Comment by danforrest** [comment]
Thanks!  Fixed the formatting.

**Comment by clintonskitson** [comment]
@danforrest Can we address the verboseness of the error when the datasubnet is not found?

```
[root@localhost temp]# docker run -ti --volume-driver=rexray -v tt1:/tt1 busybox
Error response from daemon: Cannot start container c33e35d76e43af48d1b46e5ac41b0c702c5cac5d9613aa5105843b4ff06f3353: Plugin Error: VolumeDriver.Mount, {"Error":"Missing Instance ID"}
[root@localhost temp]# vi /etc/rexray/config.ymlrexra^C
[root@localhost temp]# ^C
[root@localhost temp]# rexray volume mount ^C
[root@localhost temp]# ./rexray volume mount --volumename=tt1
FATA[0000] Missing Instance ID  
```

**Comment by danforrest** [comment]
Is it that the error is too verbose?  Not verbose enough?  Isn't helpful?

**Comment by clintonskitson** [comment]
I don't believe it is helpful.

**Comment by clintonskitson** [comment]
@danforrest LGTM

**Comment by danforrest** [comment]
Made a change in GoIsilon to fix issue #253 so I've updated glide.yaml to point to the latest version of GoIsilon.

**Comment by clintonskitson** [comment]
Gettin a `nil exception` from https://github.com/danforrest/rexray/blob/7940a7b73d4c2e98a21bd6bd710178d294bb5ba0/drivers/storage/isilon/isilon.go#L233-L237

```
[root@localhost temp]# rexray volume create --volumename=data20 --size=1
PANI[0000] runtime error: invalid memory address or nil pointer dereference 
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: (*logrus.Entry) (0x18fa760,0xc8202a4300)
[signal 0xb code=0x1 addr=0x0 pc=0x620ca5]

goroutine 1 [running]:
github.com/emccode/rexray/vendor/github.com/Sirupsen/logrus.Entry.log(0xc820056140, 0xc8202a0300, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc8202a42c0, ...)
	/home/travis/gopath/src/github.com/emccode/rexray/vendor/github.com/Sirupsen/logrus/entry.go:113 +0x618
github.com/emccode/rexray/vendor/github.com/Sirupsen/logrus.(*Entry).Panic(0xc8200ad618, 0xc8200ad700, 0x1, 0x1)
	/home/travis/gopath/src/github.com/emccode/rexray/vendor/github.com/Sirupsen/logrus/entry.go:169 +0xdf
github.com/emccode/rexray/vendor/github.com/Sirupsen/logrus.(*Logger).Panic(0xc820056140, 0xc8200ad700, 0x1, 0x1)
	/home/travis/gopath/src/github.com/emccode/rexray/vendor/github.com/Sirupsen/logrus/logger.go:157 +0xe7
github.com/emccode/rexray/vendor/github.com/Sirupsen/logrus.Panic(0xc8200ad700, 0x1, 0x1)
	/home/travis/gopath/src/github.com/emccode/rexray/vendor/github.com/Sirupsen/logrus/exported.go:107 +0x41
github.com/emccode/rexray/rexray/cli.(*CLI).Execute.func1()
	/home/travis/gopath/src/github.com/emccode/rexray/rexray/cli/cli.go:187 +0x2e0
github.com/emccode/rexray/rexray/cli.(*CLI).execute.func1()
	/home/travis/gopath/src/github.com/emccode/rexray/rexray/cli/cli.go:206 +0xd4
github.com/emccode/rexray/drivers/storage/isilon.(*driver).CreateVolume(0xc82026a280, 0xc82026f000, 0x7fff91deb7bb, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/travis/gopath/src/github.com/emccode/rexray/drivers/storage/isilon/isilon.go:237 +0x325
```

**Comment by clintonskitson** [comment]
@danforrest Can you take another peak at this one?  I would like to get your PR's merged later this week for 0.3.2.

**Comment by danforrest** [comment]
I am still not able to repro this bug.  Examining the code, the only path I can see that has a nil quota at that line also has a non-nil err so I'm really confused as to how it's producing a nil pointer there.

**Comment by clintonskitson** [comment]
Interesting. I just gave it another shot and am not hitting the same nil exception.  Looked through the code to investigate and it seems to be valid.

**Comment by clintonskitson** [comment]
@danforrest Still getting an `operation not permitted` error with latest branch when doing the chown operations.  Am I missing a setting for groups or a Isilon setting first?  Let's make sure it is documented in the PR if there is a specific setting that needs to be set in the Isilon admin console or via `config.yml`.

**Comment by danforrest** [comment]
There are about 5 privileges the user account needs (see above).  Does your test user have those privileges?  This would need to be done on the Isilon cluster.

**Comment by clintonskitson** [comment]
Ok, I see that now.  Is the user that I would set this for `root`?  Not seeing where in the interface to set these.  Is there a CLI command that can be copy pasted to make this easier or instructions to help someone configure this correctly?

**Comment by danforrest** [comment]
Set these for the user listed in config.yml.  Root should have all privileges already so if you're using root as your user, then the problem is most likely something else.  To add the privileges to another user, create a test role with the privileges and add the user to the role.  Here are the CLI commands to do it:

isi auth roles create test_role
isi auth roles modify test_role --add-user test_user
isi auth roles modify test_role --add-priv ISI_PRIV_NS_IFS_ACCESS
isi auth roles modify test_role --add-priv ISI_PRIV_NFS
isi auth roles modify test_role --add-priv ISI_PRIV_LOGIN_PAPI
isi auth roles modify test_role --add-priv ISI_PRIV_QUOTA
isi auth roles modify role_test --add-priv ISI_PRIV_IFS_RESTORE

**Comment by clintonskitson** [comment]
Excellent, works great.  Only tweak is the last line you have `role_test` reversed for the example.  I also ran into an issue relating to the group being set in my `config.yml`.  Can you update the PR with documentation relating to these settings, relation to group name, or the option to use root?

**Comment by danforrest** [comment]
What issue did you have with the group name?

I added documentation for the `group` parameter to isilon.md.  There was existing documentation addressing the required privileges.

**Comment by clintonskitson** [comment]
I guess as a complete newbie to Isilon I didn't know what it did.  I had
nobody set for troubleshooting and it conflicted with the user settings I
had.  So maybe something about what it impacts?

Basically someone should be able to get all of the features configured in
an easy way.. It sounds like that means configure everything but group and
ensure the privileges are set if not using root.

**Comment by danforrest** [comment]
Ah, gotcha.  I've updated the documentation a bit more to help make this clearer.  I added a section on optional parameters and added the names for privileges as seen on the web UI.  Let me know if you think it needs more.

**Comment by clintonskitson** [comment]
Sorry last updates.  "This would reflect a directory create under /ifs/volumes/rexray" doesn't look to be accurate, I believe it is `/ifs/volumes/`.

The "working" example should be a minimum set of parameters for best practices.  So I would remove the group from there.  I think it could use a paragraph at the top that describes how the plugin actually works.  ie. 1) creates logical volumes in directories on isilon FS 2) restricts a single host to use a volume at a time through exports 3) uses quotas to ensure that volume directory doesn't exceed size, etc..

**Comment by danforrest** [comment]
Changed to "This example would reflect a directory create under `/ifs/volumes/rexray` for created volumes."

Removed group from the example configuration.

Added a couple sentences to the over view paragraph explaining how the plugin works.

**Comment by clintonskitson** [comment]
LGTM

**Comment by clintonskitson** [comment]
@akutz let's merge this before RC1

**Pull request merged**